stm/proofBlockDelimiter.ml

```diff
@@ -1,193 +0,0 @@
-let of_vernac_control_val, to_vernac_control_val = DynBlockData.Easy.make_dyn "vernac_control"
-
-let simple_goal sigma g gs =
-  let open Evar in
-  let open Evd in
-  let open Evarutil in
-  let () = assert (not @@ Evd.is_defined sigma g) in
-  let evi = Evd.find_undefined sigma g in
-  Set.is_empty (evars_of_term sigma (Evd.evar_concl evi)) &&
-  Set.is_empty (evars_of_filtered_evar_info sigma (nf_evar_info sigma evi)) &&
-  not (List.exists (Proofview.depends_on sigma g) gs)
-
-let is_focused_goal_simple ~doc id =
-  match state_of_id ~doc id with
-  | Expired | Error _ | Valid None -> `Not
-  | Valid (Some { interp = { Vernacstate.Interp.lemmas } }) ->
-    Option.cata (Vernacstate.LemmaStack.with_top ~f:(fun proof ->
-        let proof = Declare.Proof.get proof in
-        let Proof.{ goals=focused; stack=r1; sigma } = Proof.data proof in
-        let rest = List.(flatten (map (fun (x,y) -> x @ y) r1)) @ (Evd.shelf sigma) @ (Evar.Set.elements @@ Evd.given_up sigma) in
-        if List.for_all (fun x -> simple_goal sigma x rest) focused
-        then `Simple focused
-        else `Not)) `Not lemmas
-
-type 'a until = [ `Stop | `Found of static_block_declaration | `Cont of 'a ]
-
-let crawl { entry_point; prev_node } ?(init=Some entry_point) f acc =
-  let rec crawl node acc =
-    match node with
-    | None -> None
-    | Some node ->
-        match f acc node with
-        | `Stop -> None
-        | `Found x -> Some x
-        | `Cont acc -> crawl (prev_node node) acc in
-  crawl init acc
-
-end
-
-include Util
-
-(* ****************** - foo - bar - baz *********************************** *)
-
-let static_bullet ({ entry_point; prev_node } as view) =
-  let open Vernacexpr in
-  assert (not (Vernacprop.has_query_control entry_point.ast));
-  match entry_point.ast.CAst.v.expr with
-  | VernacSynPure (VernacBullet b) ->
-      let base = entry_point.indentation in
-      let last_tac = prev_node entry_point in
-      crawl view ~init:last_tac (fun prev node ->
-        if node.indentation < base then `Stop else
-        if node.indentation > base then `Cont node else
-        if Vernacprop.has_query_control node.ast then `Stop
-        else match node.ast.CAst.v.expr with
-        | VernacSynPure (VernacBullet b') when b = b' ->
-          `Found { block_stop = entry_point.id; block_start = prev.id;
-                   dynamic_switch = node.id; carry_on_data = of_bullet_val b }
-        | _ -> `Stop) entry_point
-  | _ -> assert false
-
-let dynamic_bullet doc { dynamic_switch = id; carry_on_data = b } =
-  match is_focused_goal_simple ~doc id with
-  | `Simple focused ->
-      ValidBlock {
-         base_state = id;
-         goals_to_admit = focused;
-         recovery_command = Some (CAst.make Vernacexpr.{ control = []; attrs = []; expr = VernacSynPure (VernacBullet (to_bullet_val b))})
-      }
-  | `Not -> Leaks
-
-let () = register_proof_block_delimiter
-  "bullet" static_bullet dynamic_bullet
```

Comment:
I think this file is here for the effects like this one.

Reply:
Indeed I don't think we test that in test-suite.

I'll move this PR to draft for now; we could gain a bit from removing the error recovery thing of the STM, which AFAICT newer worked too well (at least in SerAPI / jsCoq we had to disable it, otherwise things got really flaky)

On the other hand that's not a super gain from my own experiments.

Reply:
IIRC the advantage of having the analysis of "sub proof boxes" in Coq proper rather than the document manager is that one needs the internals of the proof engine. In particular it makes sense to recover/admit a broken proof
as in ` { broken proof } .. { another proof }` only if the type of the broken proof is ground (no evar the broken proof should be filled occur in it) and if the proof term itself is not occurring in any other goal.

Maybe the code is ugly or broken, but doing it naively in the DM by putting `admit` when the structure allows for it is not sound.

Reply:
We can of course say it is not the STM that lets one skip over broken sub-proofs, but the code doing the analysis really belongs to Coq's proof model.

Reply:
What's you definition of "sound" in this context?